installation/manifest/parser.go

```diff
@@ -113,14 +114,30 @@ func (p *parser) Parse(path string, vars boshtpl.Variables, op patch.Op, release
 
 	installationManifest := Manifest{
 		Name: comboManifest.Name,
-		Template: ReleaseJobRef{
-			Name:    comboManifest.CloudProvider.Template.Name,
-			Release: comboManifest.CloudProvider.Template.Release,
-		},
 		Mbus: comboManifest.CloudProvider.Mbus,
 		Cert: comboManifest.CloudProvider.Cert,
 	}
 
+	templateList := []ReleaseJobRef{}
+	if len(comboManifest.CloudProvider.Templates) != 0 {
+		for _, template := range comboManifest.CloudProvider.Templates {
+			templateName := template.Name
+			templateList = append(templateList, ReleaseJobRef{Name: templateName, Release: template.Release})
+		}
+
+		installationManifest.Templates = templateList
+	} else {
+		templateList = append(templateList, ReleaseJobRef{
+			Name:    comboManifest.CloudProvider.Template.Name,
+			Release: comboManifest.CloudProvider.Template.Release,
+		})
+		installationManifest.Templates = templateList
+	}
+
+	// FIXME: Do we need to deduplicate the templateList? It is illegal to have multiple
+	//        entries with the same Name, but not illegal to have multiple entries with
+	//        different Names but the same Release.
+
```

Comment:
Doesn't feel necessary to me, but we certainly could. We should either do it or remove the FIXME though.

Similar to how you can't have multiple jobs with the same name, even if they come from different releases. I would expect you can't have two templates with the same name here either since I'd expect they both end up in the same folder in the "installation" path that is generated.

Reply:
Darn, I was hoping whoever reviewed this would have a strong opinion either way.

I'll add in the logic to detect duplicate templates with the same name and fail with a useful error message.

Reply:
Okay, I think I've addressed both these change requests. I pushed up a change with the (lightly-doctored) diff below:

```patch
diff --git a/installation/manifest/parser.go b/installation/manifest/parser.go
index 1cfa4ee65..4b0e08be2 100644
--- a/installation/manifest/parser.go
+++ b/installation/manifest/parser.go
@@ -119,10 +119,13 @@ func (p *parser) Parse(path string, vars boshtpl.Variables, op patch.Op, release
  }

  templateList := []ReleaseJobRef{}
+ templateNameCount := make(map[string]int)
  if len(comboManifest.CloudProvider.Templates) != 0 {
    for _, template := range comboManifest.CloudProvider.Templates {
      templateName := template.Name
      templateList = append(templateList, ReleaseJobRef{Name: templateName, Release: template.Release})
+     // Check for duplicate names
+     templateNameCount[templateName] = templateNameCount[templateName] + 1
    }

    installationManifest.Templates = templateList
@@ -134,9 +137,16 @@ func (p *parser) Parse(path string, vars boshtpl.Variables, op patch.Op, release
    installationManifest.Templates = templateList
  }

- // FIXME: Do we need to deduplicate the templateList? It is illegal to have multiple
- //        entries with the same Name, but not illegal to have multiple entries with
- //        different Names but the same Release.
+ duplicateTemplateNames := []string{}
+ for templateName, count := range templateNameCount {
+   if count > 1 {
+     duplicateTemplateNames = append(duplicateTemplateNames, "'"+templateName+"'")
+   }
+ }
+
+ if len(duplicateTemplateNames) != 0 {
+   return Manifest{}, bosherr.WrapErrorf(err, "Duplicate templates names are illegal. Duplicate template names found: %v", strings.Join(duplicateTemplateNames, ", "))
+ }

  properties, err := biproperty.BuildMap(comboManifest.CloudProvider.Properties)
  if err != nil {
  diff --git a/installation/manifest/validator.go b/installation/manifest/validator.go
index b287debc7..12934d4c4 100644
--- a/installation/manifest/validator.go
+++ b/installation/manifest/validator.go
@@ -29,7 +29,7 @@ func (v *validator) Validate(manifest Manifest, releaseSetManifest birelsetmanif

  // When there is nothing in templates, return an error. It should have a CPI release.
  if len(manifest.Templates) == 0 {
-   return fmt.Errorf("manifest.Templates cannot be empty and must contain one release")
+   return fmt.Errorf("either cloud_provider.templates or cloud_provider.template must be provided and must contain at least one release")
  }

  for _, template := range manifest.Templates {

diff --git a/installation/manifest/validator_test.go b/installation/manifest/validator_test.go
index 69757e737..b4cde5c79 100644
--- a/installation/manifest/validator_test.go
+++ b/installation/manifest/validator_test.go
@@ -61,7 +61,7 @@ var _ = Describe("Validator", func() {

      err := validator.Validate(manifest, releaseSetManifest)
      Expect(err).To(HaveOccurred())
-     Expect(err.Error()).To(ContainSubstring("manifest.Templates cannot be empty and must contain one release"))
+     Expect(err.Error()).To(ContainSubstring("either cloud_provider.templates or cloud_provider.template must be provided and must contain at least one release"))
    })

    It("validates template must be fully specified", func() {
diff --git a/installation/manifest/parser_test.go b/installation/manifest/parser_test.go
index 4e7a0b47c..bff40d628 100644
--- a/installation/manifest/parser_test.go
+++ b/installation/manifest/parser_test.go
@@ -651,6 +651,42 @@ cloud_provider:
          Mbus: "http://fake-mbus-user:fake-mbus-password@0.0.0.0:6868",
        }))
      })
+
+     Context("when there are multiple Jobs with the same names in the specified templates", func() {
+       BeforeEach(func() {
+         err := fakeFs.WriteFileString(comboManifestPath, `
+---
+name: fake-deployment-name
+cloud_provider:
+  templates:
+    - name: fake-cpi-job-name
+      release: fake-cpi-release-name
+    - name: fake-cpi-job-name
+      release: fake-cpi-release-other-name
+    - name: fake-cpi-job-name-not-duplicate
+      release: fake-cpi-release
+    - name: fake-cpi-job-name-number-two
+      release: fake-cpi-release
+    - name: fake-cpi-job-name-number-two
+      release: fake-cpi-release
+  mbus: http://fake-mbus-user:fake-mbus-password@0.0.0.0:6868
+  properties:
+    fake-property-name:
+      nested-property: fake-property-value
+`)
+         Expect(err).ToNot(HaveOccurred())
+       })
+
+       It("detects the duplicate names and returns an error containing only the duplicate names", func() {
+         _, err := parser.Parse(comboManifestPath, boshtpl.StaticVariables{}, patch.Ops{}, releaseSetManifest)
+
+         Expect(err).To(HaveOccurred())
+         Expect(err.Error()).To(ContainSubstring("'fake-cpi-job-name'"))
+         Expect(err.Error()).To(ContainSubstring("'fake-cpi-job-name-number-two'"))
+         Expect(err.Error()).ToNot(ContainSubstring("'fake-cpi-job-name-not-duplicate'"))
+       })
+
+     })
    })

    Context("when the deprecated template key is specified instead of the templates key", func() {

```